### ISSUE TITLE
Allow `TypedDict` functional syntax as base class

### ISSUE THREAD

**Issue by anniel-stripe** [python/mypy#16701]
**Bug Report**

Using `TypedDict`'s [alternative functional syntax](https://peps.python.org/pep-0589/#alternative-syntax) as a base class produces a `Unsupported dynamic base class "TypedDict" [misc]` error.

**To Reproduce**
Gist: https://mypy-play.net/?mypy=latest&python=3.12&gist=78872722658fe7bd5042cba6318b0432

**Expected Behavior**

I'd expect a call to `TypedDict` to be allowed to be used as a base class:
```python
class Params(
        TypedDict(  # This should work
            "Params",
            {"is": NotRequired["str"]},
        ),
    ):
        pass
        
a: Params = { "is": "a" }  # Produces misleading error
```

**Actual Behavior**

Instead, it produces the following errors:
```
main.py:4: error: Unsupported dynamic base class "TypedDict"  [misc]
main.py:11: error: Incompatible types in assignment (expression has type "dict[str, str]", variable has type "Params")  [assignment]
```

**Your Environment**

- Mypy version used: 1.7.1
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.12
